drivers: flash: simulator

lemrey opened the pull request:
This pull request addresses issue https://github.com/zephyrproject-rtos/zephyr/issues/7403 adding a flash driver (flash.h) implementation that writes to RAM. The driver is intended to be used mostly for testing and development. It is pulled in automatically when turning on CONFIG_FLASH_SIMULATOR through a common dts overlay file.

There are a few configuration options and features, these include:
- flash base offset
- program and erase units size
- flash wear simulation including:
-- per-unit erase cycle count 
-- per-unit read operations on worn out units
- ~~failure simulation (power loss) during read/write/erase~~ This will be renew in feature patch once we found a volunteer and get to ... the consensus
- timing simulation

The driver exports several statistics through stats.h.

nvlsianpu commented on drivers/flash/Kconfig.simulator:
```diff
@@ -0,0 +1,158 @@
+	 This setting controls the hardware failure rate during write operations in units of 1/1000.
+	 This setting is intended for testing.
+
+	config FLASH_SIMULATOR_ERASE_HW_FAILURE_RATE_PM
```
remove unwanted indent.

nvlsianpu commented on samples/drivers/flash_sim/sample.yaml:
```diff
@@ -0,0 +1,11 @@
+    harness_config:
+      type: one_line
+      regex:
+        - "All tests completed."
```
new line

lemrey replied:
You mean I should have a newline at the end of this file or a newline in the string?
It's a regexp so just a substring should be okay for a match.

nvlsianpu replied:
EOF

nvlsianpu commented on samples/drivers/flash_sim/prj.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_FLASH=y
+CONFIG_FLASH_SIMULATOR=y
+CONFIG_FLASH_PAGE_LAYOUT=y
+CONFIG_MPU_ALLOW_FLASH_WRITE=y
```
I expect it is not require for the emulation driver enabling real flash writing at all.

nvlsianpu commented on drivers/flash/flash_simulator.c:
```diff
@@ -0,0 +1,324 @@
+#include <stats.h>
+#include <string.h>
+
+#define FLASH(addr) (mock_flash + (addr)-CONFIG_FLASH_SIMULATOR_BASE_ADDR)
```
spaces around `-`

lemrey commented on drivers/flash/flash_simulator.c:
```diff
@@ -0,0 +1,324 @@
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_FAILURES
+static int toss(int failure_rate_pm)
+{
+	_rand = 1103515245 * _rand + 12345;
```
The implementation of sys_rand32_get() varies across SoC/boards.
On some boards, specifically on qemu_cortex_m3, it simply wasn't random enough since it basically returned the number of cycles (which only increases) especially in a short time frame.
This is a very simple rand() implementation from glibc that I seed using sys_rand32_get() and seems to perform decently across boards and has the advantage of not requiring a specific libc implementation.

nvlsianpu replied:
Why not use rand() form newlibc? Haw it similar or worst quality?

lemrey replied:
Mostly because we would force the application to use newlibc.
Although it might be perfectly acceptable when testing, it might be undesirable if a user wants to use this driver during development or debugging, since it would force them to use a different libc implementation.

I believe an alternative to this solution would be to tweak the sys_rand32_get() implementation for qemu_cortex_m3 (/subsys/random/rand32_timer.c). Using a multiplication instead of addition there seems to work better.

nvlsianpu commented on drivers/flash/Kconfig.simulator:
```diff
@@ -0,0 +1,158 @@
+	range 1 1000000
+
+config FLASH_SIMULATOR_ERASE_PROTECT
+	bool
```
`bool y` - should be enabled by default.

lemrey replied:
Agreed

nvlsianpu commented on drivers/flash/Kconfig.simulator:
```diff
@@ -0,0 +1,158 @@
+	range 0 1000
+	depends on FLASH_SIMULATOR_SIMULATE_FAILURES
+	help
+	 This setting controls the erase API failure rate in units of 1/1000.
```
thinks need add info what the API failure mean - that the driver refuse action right after was called, does not modify memory content and returns an error code.

lemrey replied:
I agree, I will make this text more informative.

nvlsianpu commented on drivers/flash/Kconfig.simulator:
```diff
@@ -0,0 +1,158 @@
+	range 0 1000
+	depends on FLASH_SIMULATOR_SIMULATE_FAILURES
+	help
+	 This setting controls the hardware failure rate during read operations in units of 1/1000.
```
I thinks need add info what the HW failure mean - How the error looks like in memory content and that it is a silent errro so API returns success.

lemrey replied:
Agreed, I will make this text more informative.

nvlsianpu commented on drivers/flash/Kconfig.simulator:
```diff
@@ -0,0 +1,158 @@
+	bool
+	prompt "Flash simulator"
+	default n
+	depends on FLASH
```
add dependency on MULTITHREADIN as k_wait and other system resources were used.

lemrey replied:
All right!

nvlsianpu commented on drivers/flash/flash_simulator.c:
```diff
@@ -0,0 +1,324 @@
+static int flash_init(struct device *dev)
+{
+	STATS_INIT_AND_REG(flash_sim_stats, STATS_SIZE_32, "flash_sim_stats");
+	memset(mock_flash, 0xFF, ARRAY_SIZE(mock_flash));
```
think it is good idea to allow to fill the mock flash by a pattern, like triangle or so, best if filled by noise.

lemrey replied:
That's a good idea. I plan to let the user configure two separate byte patterns:
- one to be used for hardware failures and 
- one to be used when an erase unit has reached its erase cycles limit

lemrey replied:
Oh, I see now you meant to fill the flash with a pattern at initialization..
I think the same could be achieved by writing the flash through the API maybe?

nvlsianpu reviewed (changes requested):
1st of my part of the CR iteration  is done.

lemrey commented on drivers/flash/flash_simulator.c:
```diff
@@ -0,0 +1,324 @@
+static int flash_sim_read(struct device *dev, off_t offset, void *data,
+			  size_t len)
+{
+	ARG_UNUSED(dev);
```
I believe if the device is not correct there are no chances to end up in this function..
How do you suggest I check that?

nvlsianpu replied:
you are right, lets ignore my comment.

lemrey commented on samples/drivers/flash_sim/src/main.c:
```diff
@@ -0,0 +1,137 @@
+	sim_stats = stats_group_find("flash_sim_stats");
+	if (sim_stats) {
+
+		stats_walk(sim_stats, flasm_sim_stats_walk_fn, NULL);
```
This is a typo, must be fixed.

pizi-nordic commented on drivers/flash/Kconfig.simulator:
```diff
@@ -0,0 +1,181 @@
+
+if FLASH_SIMULATOR
+
+config FLASH_SIMULATOR_BASE_ADDR
```
Are we going to simulate internal flash (the one which can be mapped into CPU address space) ?
If no, this is not needed (flash will always start from 0). If yes, we should be able to detect accidental write to flash area (this can be done using MPU).

d3zd3z commented on drivers/flash/flash_simulator.c:
```diff
@@ -0,0 +1,344 @@
+static int toss(int failure_rate_pm)
+{
+	rand_seed = 1103515245 * rand_seed + 12345;
+	return ((u32_t)rand_seed % 1000) < failure_rate_pm ? -EIO : 0;
```
Not sure how much this matters, but the division is pretty expensive on some platforms.

nvlsianpu replied:
Probably no matters as we mainly intended to us it for <i>qemu</i> simulations test.

nvlsianpu replied:
hw error sim is not in this patch anymore

d3zd3z commented on drivers/flash/flash_simulator.c:
```diff
@@ -0,0 +1,344 @@
+	STATS_INCN(flash_sim_stats, bytes_written, len);
+	for (u32_t i = 0; i < len; i++) {
+		/* only pull bits to zero */
+		*(FLASH(offset + i)) &= *((u8_t *)data + i);
```
This is not how most flash devices these days work. For the most part, flash devices have small writable units (typically 1-8 bytes), and these sections must only be written to once. On some devices, even writing 0xFF and then trying to write something else will fail. Other devices don't fail, but the data sheets suggest that it greatly increases wear on the devices.

I would suggest that we consider writes to anything that isn't all 0xFF to be a failure.

We might be interested, also, in enforcing a particular alignment of the writes.

lemrey replied:
You have a point, still I would make this behavior configurable since some hardware does allow multiple writes to the same address/region.

aurel32 replied:
it's true that some flash devices can only be written once, usually (but not exclusively) the ones with ECC. This is the case for example of the STM32L4 or STM32H7. On the other hand some of them like the STM32F1 support being written multiple time, and it's even a trick often used to implement a chip (down)counter, changing one bit from 1 to 0 at each counter (dec)increment.

In short I agree with @lemrey that it should be a configurable behavior. Maybe it should not be enabled by default as most of new hardware released recently do not allow multiple write.

I also agree with @nvlsianpu that adding support for write block size would be nice to have. I guess @Laczen can confirm that it would have been useful to test NVS.

nvlsianpu replied:
@Laczen as @lemrey  said failure modes configuration allows  to simulate  failures expected by nor-flash devices. For now  we want to us this driver for testing stuff without failure emulation.

nvlsianpu replied:
That said I mean that we need to start using this, and then we can improve if this will be a case.

d3zd3z reviewed (dismissed):
So, there is a lot of code here to simulate "failures", but I don't think these are really the typical ways that NOR flash devices fail. By far, the most common failure with flash operations is to have power removed while performing the operation. This can leave both erases and writes partially done, resulting in bits that read different values, depending on things like temperature.

The next failure is going to be failure from device wear. This generally results in erases that don't complete.

Failures where just wrong values are written are very rare, and it is unlikely (and frankly unreasonable) to expect the client of the flash to be able to handle this kind of failure. I don't see a strong reason to test this case. If the flash device in question is indeed unreliable (typical for NAND devices), it will need ECC in order to correct the errors, not to expect a filesystem to be able to handle the errors.

lemrey commented:
Hi @d3zd3z, thank you for your comments. The "hardware failures" here are really meant to simulate power loss during write and erase operations, and read errors due to wear, rather than to simulate a defective hardware. Perhaps I have used a misleading terminology? :)

nvlsianpu commented:
@lemrey can you resolve conflicts?

nvlsianpu commented:
@carlescufi This PR is long time in the queue - can we merge it?

carlescufi commented:
@nvlsianpu let's get an update from the corresponding parties.

@lemrey is this still applicable and if so can you rebase it to the latest master please?
@d3zd3z can you please revisit this?
@aurel32 @Laczen Could you take a look at this please?

lemrey commented:
@carlescufi yes it is still applicable, and I rebased it just this morning.

nvlsianpu commented:
@carlescufi Is this looks like silent approval?

carlescufi commented:
@Laczen @d3zd3z @aurel32 Please re-review

d3zd3z commented:
I still don't think this is really simulating flash failures in any way that an actual flash device would fail. The `CONFIG_FLASH_SIMULATOR_SIMULATE_FAILURES` seems to toggle all of the bits randomly on each read. When a device (ordinary NOR) does wear out, usually the erase will fail first. For the most part, you'll actually get an error when doing the erase, and can handle this in software. In either case, reads wouldn't just change some random byte somewhere, but the same area would be written incorrectly. The simulated failure is something I wouldn't consider reasonable for any software to handle. Since I wouldn't expect to see it.

If you want to simulate real device failures, you need to have some mechanism for interrupting some particular operation, and then instead of returning, restart the system. This gives whatever software is using the flash a chance to execute its recovery code. It's usually pretty hard to do this on target (which is why the MCUboot simulator runs off-target). If you can make the simulated flash memory uninitialized and it survives reset, just doing resets is probably the most realistic.

A partial erase will result in the data reading back as, well, partially erased, with some of the bits set to one. A partial write will result in the last `PROG_UNIT` bytes not being fully written (meaning it will have more 1 bits than the desired value). This should be done at write time, into the simulated buffer, and not really at read time.

nvlsianpu commented:
@lemrey Are you able to adjust patch to @d3zd3z suggestions? Let me know especially if you refuse.

nvlsianpu commented:
@carlescufi
So we decided to limit this patch to simulate ideal flash driver only - we resign form submission the failure simulator and the example as it doesn't mak any sense for regular driver operation.

Except the above the code kept here has the same shape as it has before.

nashif reviewed (dismissed):
Can we please have a testcase that uses this new driver?

nashif commented:
ping

nvlsianpu commented on drivers/flash/Kconfig.simulator:
```diff
@@ -0,0 +1,95 @@
+	help
+	 The smallest area of flash memory that can be indipendently programmed.
+
+config FLASH_SIMULATOR_FLASH_SIZE
```
I think not to put this to DT as this is the simulator.  Interaction with Kconfig are much easier than with DT, and theirs a GUI for it.

kapi-no replied:
Moved this part of configuration to DT

nvlsianpu commented:
recheck

nvlsianpu commented:
@lemrey, @nashif Can you look at changes  - I added test, reorganized DT patch (as DT shall not depends on Kconfig) and of course got rid out a few bugs.

nvlsianpu commented:
recheck

nvlsianpu commented:
cc @carlescufi ^^

nvlsianpu commented:
@lemrey `The last commit `drivers: flash: fix flash-simulator build crash`  fixes build isseu you have observing before.

nvlsianpu commented:
recheck

nvlsianpu commented:
recheck

lemrey commented on drivers/flash/flash_simulator.c:
```diff
@@ -0,0 +1,283 @@
+#if (defined(CONFIG_STATS) && (CONFIG_FLASH_SIMULATOR_FLASH_SIZE > 40))
+/* Check here as it is imposible to make such a limitation in Kconfig */
+#error You shall not declare more than 40 pages while using statistic
+	/* Imitation above is caused by used UTIL_REPEAT                   */
```
nitpick: Imitation -> Limitation

lemrey reviewed (commented):
LGTM, thanks for following this through.

nvlsianpu commented:
recheck

nvlsianpu commented:
recheck

nvlsianpu commented:
recheck

carlescufi commented:
@lemrey can you address the current questions and fix the build so that we can consider merging this?

kapi-no commented:
@carlescufi, I took over the maintenance of this PR.

I resolved conflicts, addressed some remaining comments and I'm waiting for CI so that this PR is mergable again.

kapi-no commented on drivers/flash/flash_simulator.c:
```diff
@@ -0,0 +1,273 @@
+{
+	STATS_INIT_AND_REG(flash_sim_stats, STATS_SIZE_32, "flash_sim_stats");
+	memset(mock_flash, 0xFF, ARRAY_SIZE(mock_flash));
+	rand_seed = sys_rand32_get();
```
removed this leftover

lemrey commented on drivers/flash/flash_simulator.c:
```diff
@@ -0,0 +1,273 @@
+	return 0;
+}
+
+DEVICE_AND_API_INIT(flash_simulator, "FLASH_SIMULATOR", flash_init, NULL, NULL,
```
Yes, this wasn't considered initially but could be added. I don't see much value in doing that though. This is a simulator and in practice it will be used to test libraries/subsystems, in which case one flash device is probably enough.

nashif reviewed (changes requested):
Please remove the changes to nordic dtsi, they seem unrelated to the change. 
Can we use this to run a few of the existing tests involving flash in qemu? for example flash_map? It would be great to see an example of how this will work with existing tests.

nvlsianpu commented on drivers/flash/Kconfig.simulator:
```diff
@@ -0,0 +1,95 @@
+	depends on FLASH
+	select STATS
+	select STATS_NAMES
+	select MULTITHREADING
```
^^ MULTITHREADING

kapi-no committed:
bluetooth: host: format specifier alignment

Aligned format specifiers for the Bluetooth Host. Now, the format
specifier matches the variable type.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no commented:
@nashif:
- removed the changes to nordic dtsi
- added qemu platform to the flash map test
- addressed other review comments

nvlsianpu commented on drivers/flash/Kconfig.simulator:
```diff
@@ -0,0 +1,69 @@
+
+if FLASH_SIMULATOR
+
+config FLASH_SIMULATOR_FLASH_SIZE
```
FLASH_SIMULATOR_FLASH_SIZE should be replaced by register size in flash_sim0: node you introduced. (https://github.com/zephyrproject-rtos/zephyr/blob/4952168c14c06e990489de927482155c44665e97/dts/common/sim-flash.dtsi#L15)

nvlsianpu reviewed (changes requested):
more comments tomorrow

nvlsianpu commented on boards/arm/qemu_cortex_m3/qemu_cortex_m3.dts:
```diff
@@ -69,3 +71,30 @@
+		/* Storage partition will be used by FCB/NFFS/NVS if enabled. */
+		storage_partition: partition@3000 {
+			label = "storage";
+			reg = <0x00003000 0x00001000>;
```
set size to 0x10000. this will be enought for most of tests.

nvlsianpu commented on boards/arm/qemu_cortex_m3/qemu_cortex_m3.dts:
```diff
@@ -69,3 +71,30 @@
+
+		slot0_partition: partition@1000 {
+			label = "image-0";
+			reg = <0x00001000 0x00001000>;
```
set size of slot_0/1 to some bigger size like 32 KB.

nvlsianpu reviewed (changes requested):
* add me or you to the codeowenr file for this subsystem and its test
* add missing license clause where needed
* fix gitlint issues
* fix kconfig settings for flash_map test on qemu
```
warning: MPU_ALLOW_FLASH_WRITE (defined at arch/arm/core/cortex_m/mpu/Kconfig:41) was assigned the
value 'y' but got the value 'n'
```
MPU_ALLOW_FLASH_WRITE is not avielable for qemu.

kapi-no committed:
drivers: flash: add flash_simulator driver

This commit adds a flash driver implementation that writes to RAM and
exports statistics through stats.h. It can be used to simulate flash
memory for testing purposes.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>
Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
tests: drivers: add flash_simulator driver

This commit adds a tests for flash simulator driver implementation
for qemu_x86 boards.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
drivers: flash: fix flash-simulator build crash

Statistic for flash has limitation regards number of pages possible to
be counted. This path introduces check for that in source file as it is
not possible to preserve the limit in Kconfig.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
boards: changing flash simulator device name alias

Changed alias used for flash simulator device name to match the real
one.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
drivers: flash: flash_simulator enabled by default on qemu

Enabled Flash Simulator on the QEMU x86 platform by default.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
tests: storage: flash_map: adding qemu platform

Added QEMU platform to the Flash Map test and defined partitions for
QEMU dts.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
boards: qemu_x86: reducing flash size

Reduced flash size for QEMU x86 so it does not consume a lot of RAM
memory where it is simulated.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
driver: flash: flash_simulator: moved part of Kconfig to DT

Moved part of the KConfig configuration to the Device Tree.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no commented:
Migrated from qemu_cortex_m3 to qemu_x86, so it's possible to configure partitions according to @nvlsianpu recommendations (more RAM for simulated Flash array).

Please recheck

nvlsianpu reviewed (approved):
Thanks for your work Kamil.

galak commented:
There are a number of device tree issues that need cleaning up in this PR.  How do we want to handle that?

nvlsianpu commented:
what issues?

galak commented:
1. use of #ifdef in the dts files (XIP)
2. naming of the dts compat
3. adding dts/common/sim-flash.dtsi
4. Use of CONFIG_ defines for non-Kconfig symbols
5. referencing board Kconfig in drivers (not something we typically do)

galak commented on CODEOWNERS:
```diff
@@ -158,6 +158,7 @@
 /dts/bindings/*/st,stm32*                 @erwango
 /dts/bindings/sensor/ams,ens210.yaml      @alexanderwachter
 /dts/bindings/sensor/ams,iaqcore.yaml     @alexanderwachter
+/dts/common/sim-flash.dtsi                @nvlsianpu
```
lets remove sim-flash.dtsi and just put the flash node in the qemu dts

galak commented on boards/x86/qemu_x86/qemu_x86.dts:
```diff
@@ -13,6 +13,7 @@
 #endif
 
 #include <ia32.dtsi>
+#include "../../../dts/common/sim-flash.dtsi"
```
just inline the flash node.

galak commented on drivers/flash/Kconfig.simulator:
```diff
@@ -0,0 +1,61 @@
+menuconfig FLASH_SIMULATOR
+	bool
+	prompt "Flash simulator"
+	default y if BOARD_QEMU_X86
```
We should put this in the board not here.

galak commented on dts/bindings/flash_controller/sim-flash.yaml:
```diff
@@ -0,0 +1,22 @@
+      type: string
+      category: required
+      description: compatible strings
+      constraint: "sim-flash"
```
This should be "zephyr,sim-flash"

kapi-no replied:
I will also rename the file as it seems that this constraint value is aligned with the yaml file name in other examples

galak commented on tests/drivers/flash_simulator/src/main.c:
```diff
@@ -0,0 +1,276 @@
+#include <device.h>
+
+/* configuration derived from DT */
+#define CONFIG_FLASH_SIMULATOR_BASE_OFFSET DT_FLASH_BASE_ADDRESS
```
These should not be `CONFIG_` prefix since these aren't related to any Kconfig options.

galak reviewed (commented):
Various DT changes / code changes.

kapi-no commented:
> There are a number of device tree issues that need cleaning up in this PR. How do we want to handle that?

@galak, I will address these issues soon in a separate PR.

Vudentz commented:
Btw, this one breaks with tests/bluetooth/shell:

uart:~$ bt init

***** Kernel Panic! *****
Current thread ID = 0x004000e0 (shell_uart)
eax: 0xffffffea, ebx: 0x00000000, ecx: 0x00516000, edx: 0x00516014
esi: 0x0001fb3c, edi: 0x00400414, ebp: 0x00506608, esp: 0x005065f8
eflags: 0x00000282 cs: 0x0008
call trace:
eip: 0x0000a4ff
     0x0000a53c (0x50661c)
     0x0000e3e9 (0x5161b4)
     0x00011aec (0x28740)
     0x0001f536 (0x1fb3c)
     0x000063f6 (0x28740)
     0x0000659a (0xbc19)
     0x00005e3a (0x28740)
     0x000071de (0x0)
Terminate emulator due to fatal kernel error
make[3]: *** [zephyr/CMakeFiles/run.dir/build.make:60: zephyr/CMakeFiles/run] Error 1
make[2]: *** [CMakeFiles/Makefile2:335: zephyr/CMakeFiles/run.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:342: zephyr/CMakeFiles/run.dir/rule] Error 2
make: *** [Makefile:196: run] Error 2
addr2line -e zephyr/zephyr.elf 0x0000a4ff
/home/vudentz/git/zephyr-github/subsys/settings/src/settings_init.c:107

Disabling BT_SETTINGS makes it work again.
